### PR TITLE
Module3

### PR DESCRIPTION
**Pull request by uhernando002**
I have finished the milestone 1, through a notebook, where I made some inference about the problem and the models that have been trained. These are the final insights:
- The results are very similar in the 3 models. The recall obtained is of 0.28 which means that out of all positives the model will predict 28% of them. This is a boost of 28% over the selected items, which was one of the targets of the problem.
- Since the dataset is large, the regularization doesn't affect, because the overfitting is not a real problem.
- However, I think is better to choose the lasso, because there are a few variables (explained before) that have impact over the prediction. Because of that, and since the lasso model gets rid of some variables getting their betas to 0, the model will be lighter and will have less computational load.

Tomorrow I will start with milestone 2, but I wanted to check if the coding and the insigthts are adecuate for the problem.

### CODE REVIEW

**Comment by pabaldonedo** [comment]
I'd personally advocate to avoid the inplace method. Doing stuff inplace in general may unwillingly affect third parties that are pointing to the same object

_Respond and view the context [here](https://gitnotebooks.com/uhernando002/zrive-ds/pull/3#comment-579)._

**Comment by pabaldonedo** [comment]
Minor: You are not respecting PEP8

_Respond and view the context [here](https://gitnotebooks.com/uhernando002/zrive-ds/pull/3#comment-580)._

**Comment by uhernando002** [comment]
Is it better with apply? Or how must be done in this cases?

_Respond and view the context [here](https://gitnotebooks.com/uhernando002/zrive-ds/pull/3#comment-581)._

**Comment by pabaldonedo** [comment]
Good point.
For the split is very common to use [this sklearn method](https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.train_test_split.html)

_Respond and view the context [here](https://gitnotebooks.com/uhernando002/zrive-ds/pull/3#comment-582)._

**Comment by pabaldonedo** [comment]
you could just bb_df.loc[:, "product_type"] = bb_df.product_type.replace(((bb_df['product_type'].value_counts()/len(bb_df)*100).round(5)).to_dict(), inplace=True)

_Respond and view the context [here](https://gitnotebooks.com/uhernando002/zrive-ds/pull/3#comment-583)._

**Comment by uhernando002** [comment]
But if I need validation? I tried that method, but I thought it only divide the dataset into train and test, not validation. I thought that this method was better in this case

_Respond and view the context [here](https://gitnotebooks.com/uhernando002/zrive-ds/pull/3#comment-584)._

**Comment by uhernando002** [comment]
Okey, great

_Respond and view the context [here](https://gitnotebooks.com/uhernando002/zrive-ds/pull/3#comment-585)._

**Comment by pabaldonedo** [comment]
Here you have made a severe mistake on data leakage. The categorical variables have computed using the entire dataset prior to the split. There are couple of ways to avoid this:
1. Quick:
  - Make the train-test split
  - Compute the encoding in train
  - Apply the encoding to val & test

2. Better but takest more time:
  - You could create a class that makes the transformation that extends TransformerMixin and BaseEstimator classes of sklearn
  - In your model, you create a [Pipeline](https://scikit-learn.org/stable/modules/classes.html#module-sklearn.pipeline) made of ColumnTransformer followed by the actual model
  - Apply train to the train and predict to the val and test
  - A talk on this [here](https://www.youtube.com/watch?v=BFaadIqWlAg)
  - This is an advance topic so do not go for it unless you have plenty of spare time and feel like learning about it. This is a scikit-learn related stuff and not the core of what we should be learning this week

_Respond and view the context [here](https://gitnotebooks.com/uhernando002/zrive-ds/pull/3#comment-588)._

**Comment by pabaldonedo** [comment]
Minor: It is common practice to keep the imports all at the top of the notebook

_Respond and view the context [here](https://gitnotebooks.com/uhernando002/zrive-ds/pull/3#comment-589)._

**Comment by pabaldonedo** [comment]
interesting choice of the solver. How did you decide the max_iter?

_Respond and view the context [here](https://gitnotebooks.com/uhernando002/zrive-ds/pull/3#comment-590)._

**Comment by pabaldonedo** [comment]
It seems that your solution has not converged (look at the warning), what was your reaction to that warning? What were your steps after seeing it? Did you simply ignore it?

_Respond and view the context [here](https://gitnotebooks.com/uhernando002/zrive-ds/pull/3#comment-591)._

**Comment by pabaldonedo** [comment]
Is it "Logical" the positive relation of abanadoned before? What are your thoughts on that?

_Respond and view the context [here](https://gitnotebooks.com/uhernando002/zrive-ds/pull/3#comment-592)._

**Comment by uhernando002** [comment]
IT was a bit random. With 100 iterations it appeard a warning that said that the model didn´t reach the minimum error. However, with 250 also appeard the warning, but I imagine that fits better. What do you think about that?

_Respond and view the context [here](https://gitnotebooks.com/uhernando002/zrive-ds/pull/3#comment-593)._

**Comment by pabaldonedo** [comment]
I like your reasoning. Here are some challenges you could face:
- What if push notifications are carried out by a third party that charge us per push, would FP then be neglible?
- UX is important, excess of pushes may lead to user uninstalling our "disturbing" app, what is the cost of FPs that lead to this situation?

_Respond and view the context [here](https://gitnotebooks.com/uhernando002/zrive-ds/pull/3#comment-594)._

**Comment by uhernando002** [comment]
I only increased the max iterations. But continued the error. I don´t really know why doesn´t converge :(

_Respond and view the context [here](https://gitnotebooks.com/uhernando002/zrive-ds/pull/3#comment-595)._

**Comment by pabaldonedo** [comment]
I like your initiative and you will surely face situations when you will have to make the call for this. Nevertheless, bear in mind that here probably a discussion with the product team would be necessary to have the full information of the costs of FPs and FNs

_Respond and view the context [here](https://gitnotebooks.com/uhernando002/zrive-ds/pull/3#comment-596)._

**Comment by uhernando002** [comment]
It can be logical or illogical depending on how you look at it. 
- If you take abandoned as a time expiration, where the user went to do other things and left the request uncompleted, it is logical that the relationship is positive, because the fact of abandonment in these cases means that they most likely finish executing the buys. 
- On the contrary, if abandonment is taken to mean that the user is no longer interested in the cart for x reason, the relationship should be negative. 
Therefore, it should be concluded that logical reasoning is the first, given the relationship that the model presents between the variables.

_Respond and view the context [here](https://gitnotebooks.com/uhernando002/zrive-ds/pull/3#comment-597)._

**Comment by pabaldonedo** [comment]
Being super picky here: around 10% of male population has a sort of daltonism. Probably this palete is not the best. [This](https://scikit-learn.org/stable/auto_examples/model_selection/plot_confusion_matrix.html#sphx-glr-auto-examples-model-selection-plot-confusion-matrix-py) palette for example suits better in my opinion

_Respond and view the context [here](https://gitnotebooks.com/uhernando002/zrive-ds/pull/3#comment-598)._

**Comment by pabaldonedo** [comment]
same warnings

_Respond and view the context [here](https://gitnotebooks.com/uhernando002/zrive-ds/pull/3#comment-599)._

**Comment by pabaldonedo** [comment]
avoid variables in spanish

_Respond and view the context [here](https://gitnotebooks.com/uhernando002/zrive-ds/pull/3#comment-600)._

**Comment by pabaldonedo** [comment]
I'd opt to plot alpha against the metric of itnerest to observe how the model underfits/overfits as we change the regularisation parameter. With a plot we see if we have covered all the interest range of values or maybe if it is worth to explore larger or smaller values outside the range we used.

_Respond and view the context [here](https://gitnotebooks.com/uhernando002/zrive-ds/pull/3#comment-601)._

**Comment by pabaldonedo** [comment]
You are repeating a lot of code. Consider to use functions, it will make the code easier to read, more compact and less error prone. A notebook is not an excuse to be less dilligent in programming. We might relax stuff like docstirngs etc, but good practices are always helpful.

_Respond and view the context [here](https://gitnotebooks.com/uhernando002/zrive-ds/pull/3#comment-602)._

**Comment by pabaldonedo** [comment]
The leading "0" is innecessary in the index

Why are you excluding the last value?

_Respond and view the context [here](https://gitnotebooks.com/uhernando002/zrive-ds/pull/3#comment-603)._

**Comment by uhernando002** [comment]
- In that case we should reconsider the solution. We should talk with production  and business areas to see if the extra pushes that don´t carry any purchase will be compensated by the product that with a 30% more or less will be purchased thanks to the push.
- Yes that's true. But considering the relation in the Confusion Matrix, the number of FP are not too high in comparison with the TP that will be beneficial for the company. I think that the relation presented by the solution is not so bad for the UX, as there will always be a trade off between the FN and FP. What do you think?

_Respond and view the context [here](https://gitnotebooks.com/uhernando002/zrive-ds/pull/3#comment-604)._

**Comment by pabaldonedo** [comment]
It is very hard to compare the models as they are to get the insights. Once you get the best parametrisation it would have been helpful to plot the Precision-recall curves of the 3 models in the same plot, with a different color each. In that manner we could easily compare the models

_Respond and view the context [here](https://gitnotebooks.com/uhernando002/zrive-ds/pull/3#comment-605)._

**Comment by uhernando002** [comment]
Yes, absolutely

_Respond and view the context [here](https://gitnotebooks.com/uhernando002/zrive-ds/pull/3#comment-606)._

**Comment by pabaldonedo** [comment]
IS this plot correct? I think you have not execut this with LAsso since the plots look veryt similar to the ridge (And the tilte states ridge)

_Respond and view the context [here](https://gitnotebooks.com/uhernando002/zrive-ds/pull/3#comment-607)._

**Comment by pabaldonedo** [comment]
Good

_Respond and view the context [here](https://gitnotebooks.com/uhernando002/zrive-ds/pull/3#comment-608)._

**Comment by uhernando002** [comment]
Because the threshold array was 1 value smaller than the two others. (I don't know why)

_Respond and view the context [here](https://gitnotebooks.com/uhernando002/zrive-ds/pull/3#comment-609)._

**Comment by pabaldonedo** [comment]
Are this result good? What do you think?

_Respond and view the context [here](https://gitnotebooks.com/uhernando002/zrive-ds/pull/3#comment-610)._

**Comment by uhernando002** [comment]
Sure. The error rate is very similar to validation, so we didn't overfitted the model by a random and favourable dataset of validation

_Respond and view the context [here](https://gitnotebooks.com/uhernando002/zrive-ds/pull/3#comment-611)._

**Comment by pabaldonedo** [comment]
No problem by using this method. I was just mentioning in case you were unaware of it. I have seen people that simply calls the trainest split twice in this case, although I'd advocate for something like your approach tbh

_Respond and view the context [here](https://gitnotebooks.com/uhernando002/zrive-ds/pull/3#comment-612)._

**Comment by uhernando002** [comment]
It could be, as I had some problems with the variables. I will check it again

_Respond and view the context [here](https://gitnotebooks.com/uhernando002/zrive-ds/pull/3#comment-613)._

**Comment by uhernando002** [comment]
Okey

_Respond and view the context [here](https://gitnotebooks.com/uhernando002/zrive-ds/pull/3#comment-614)._

**Comment by pabaldonedo** [comment]
Have you tried L-bfgs or another one?

The warning basically means that the optimiser is an iterative method: it iterates towards the solution. At each iteration the error decreases. The method stops when the error decrease is very small. It has not reached that point, so I would not be sure that the fit is properly done

_Respond and view the context [here](https://gitnotebooks.com/uhernando002/zrive-ds/pull/3#comment-615)._

**Comment by pabaldonedo** [comment]
Do you understand how SAG works? Probably not unless you have look it up (honestly I should look it up too). My first reaction would be to try other optimisers first or increase the iterations even further unless the time it ates for training is execevily big

_Respond and view the context [here](https://gitnotebooks.com/uhernando002/zrive-ds/pull/3#comment-616)._

**Comment by pabaldonedo** [comment]
Good. I'd add that maybe it was too expensive so although it was interested it took it off.

The last sentence you should be careful, if collinearity is present (something to check if we want to do proper analysis of coefficients) it could be causing that a negative relationship has a positive coefficient in our model

_Respond and view the context [here](https://gitnotebooks.com/uhernando002/zrive-ds/pull/3#comment-617)._

**Comment by pabaldonedo** [comment]
My point here is to make you think that there can be multiple effects and causistics when we make decisions. I'd try to talk to the product guys to get more insights. As an exercise your reasoning seems good :)

_Respond and view the context [here](https://gitnotebooks.com/uhernando002/zrive-ds/pull/3#comment-618)._

**Comment by uhernando002** [comment]
Yes, I tried. But I checked in some page that with large datasets was better option the sag. 
Yes I know that it seems to not get that error smaller. But making more iterations could help to achieve that error or is simply a bad idea because it could overfit? Because in my opinion 100 iterations in this case should be sufficient

_Respond and view the context [here](https://gitnotebooks.com/uhernando002/zrive-ds/pull/3#comment-619)._

**Comment by uhernando002** [comment]
Okey, I'll try with other optimisers, and if it doens't work I'll make greater the iterations

_Respond and view the context [here](https://gitnotebooks.com/uhernando002/zrive-ds/pull/3#comment-620)._

**Comment by pabaldonedo** [comment]
If you have such doubts always refer to the [documentation](https://scikit-learn.org/stable/modules/generated/sklearn.metrics.precision_recall_curve.html). By reading carefully, under the return section you will see why that happens. Try never to have such doubts, as it may indicate a lack of understanding or a symptom of a bug.

_Respond and view the context [here](https://gitnotebooks.com/uhernando002/zrive-ds/pull/3#comment-621)._

**Comment by uhernando002** [comment]
I don't understand that negative relationship that you say, because in this case the beta is positive, what it means that with all the variables constant, the independent variable increases the beta with an increase of 1 of the variable no?

_Respond and view the context [here](https://gitnotebooks.com/uhernando002/zrive-ds/pull/3#comment-622)._

**Comment by uhernando002** [comment]
Sorry, I didn't read the documentation properly

_Respond and view the context [here](https://gitnotebooks.com/uhernando002/zrive-ds/pull/3#comment-623)._

**Comment by pabaldonedo** [comment]
1. This has nothing to do with overfitting (well one could argue yes but in very specific situations like neural networks, I'd say it does not apply here)
2. This solver is trying to find the solution (i.e. the model weights) that minimise your target function (error + regularisation)
3. We would like to have the optimum solution (optimum weights) for the current parametrisation
4. Therefore, we want this to converge.

This is quite important and may invalidate the entire analysis since you may not be finding the optimum parameters that you are seeking.

Finally, let me challenge, why do you think that 100 iterations are enough? Why not 10 or 1000?

_Respond and view the context [here](https://gitnotebooks.com/uhernando002/zrive-ds/pull/3#comment-624)._

**Comment by pabaldonedo** [comment]
good. But it is important that you understand what is the problem we are facing here since it is quite important.

_Respond and view the context [here](https://gitnotebooks.com/uhernando002/zrive-ds/pull/3#comment-625)._

**Comment by pabaldonedo** [comment]
I am refering to the general case. If you have collinearity, i.e. input variables are highly correlated, and you fit a linear model, you may end up with deceiving coefficients in what it relates to causality. You could end up with positive coefficients for a variable that actually negatively affects the output. This could happen if there is another input variable that is highly correlated to it. If you are interested in going deeper you could have a look [here](https://bookdown.org/rwnahhas/RMPH/mlr-collinearity.html) for example. But I'd focus first in finishing the exercise and only refer to this bibliography that I share with you if you have spare time and interest.

_Respond and view the context [here](https://gitnotebooks.com/uhernando002/zrive-ds/pull/3#comment-626)._

File: tests/module_2/test_2_groceries.py
Comment (pabaldonedo) [inline tests/module_2/test_2_groceries.py:1]: Why are you changing something form module 2 here?

**Comment by uhernando002** [comment]
I don't really now why not 1000. May be 10 is small to reach the convergence as the number of calculations of the gradients are small and only with few variables could be reached. However, a 100 is a good option, as it is strange to have models trained with too many variables. 
I am not sure about the answer, but it is what my reasoning says

_Respond and view the context [here](https://gitnotebooks.com/uhernando002/zrive-ds/pull/3#comment-627)._

**Comment by pabaldonedo** [comment]
My point is that we don't have any evidences to say what is a reasonable amount of iterations. I would not be able to say if 100 are enough, too few or already a lot

_Respond and view the context [here](https://gitnotebooks.com/uhernando002/zrive-ds/pull/3#comment-628)._

**Comment by uhernando002** [comment]
Okey, I understand what you say

_Respond and view the context [here](https://gitnotebooks.com/uhernando002/zrive-ds/pull/3#comment-629)._

**Comment by uhernando002** [comment]
Okey

_Respond and view the context [here](https://gitnotebooks.com/uhernando002/zrive-ds/pull/3#comment-630)._

**Comment by pabaldonedo** [comment]
Those are all very valid points. However, let's step back to the broader picture: what is the ultimate goal? To bring value to the business. We are proposing here to use a "complex" module as it is a model to bring value. Is it worth it? What if we just use a simple rule like the probability of buying is equal to the global popularity of the item? That is super simple. Are we beating that? And if we are, how much value are we bringing to the table comparing to this simple rules by hand?

_Respond and view the context [here](https://gitnotebooks.com/uhernando002/zrive-ds/pull/3#comment-631)._

**Review by pabaldonedo — COMMENTED** [review]
I like the general idea of the notebook, it is clear that you are learning the key concepts. Good work!

However, there are things yet to improve:

1. The data leakage was a severe mistake. 
2. The model not converging was also severe. Try always to understand what you are using. You read that the chose optimiser was the best for small datasets. Why? Are there any caveats? Always ask yourself if you understand what you are doing and the caveats.
3. I have left you some comments to challenging your statements. I think it is a good exercise to never lose the real focus: bring value to the company.

You should revisit the data leakage and the model training. Understand what you did wrong and why. Any questions I am happy to help.

Keep up the good work and the learning! :)

**Comment by pabaldonedo** [comment]
I completely forgot yesterday: try to apply standard scaler to the input. It not only helps with the coefficients interpretation, but also tend to help in the optimisation process

_Respond and view the context [here](https://gitnotebooks.com/uhernando002/zrive-ds/pull/3#comment-632)._

**Comment by uhernando002** [comment]
done

_Respond and view the context [here](https://gitnotebooks.com/uhernando002/zrive-ds/pull/3#comment-693)._

**Comment by uhernando002** [comment]
Done

_Respond and view the context [here](https://gitnotebooks.com/uhernando002/zrive-ds/pull/3#comment-694)._

File: src/module_3/groceries_model.pkl
Comment (pabaldonedo) [inline src/module_3/groceries_model.pkl:1]: As a general rule models, as well as data, are not incorporated into the repostiory

File: src/module_3/module_3_milestone2.py [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+import numpy as np
+from sklearn import linear_model
+from sklearn.pipeline import make_pipeline
+from sklearn.preprocessing import StandardScaler
+from sklearn import metrics
+import joblib
+
+
+DATA_PATH = "/home/unai/datasets/feature_frame.csv"
+training_columns = [
```
Comment (pabaldonedo) [inline src/module_3/module_3_milestone2.py:11]: Nice to be explicit with the columns!

File: src/module_3/module_3_milestone2.py [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+    "abandoned_before",
+    "ordered_before",
+    "global_popularity",
+    "user_order_seq",
+    "normalised_price",
+]
+label_col = "outcome"
+model_filename = "groceries_model.pkl"
```
Comment (pabaldonedo) [inline src/module_3/module_3_milestone2.py:19]: Generally, "global" variables (these are not global being rigurous) are in caps

File: src/module_3/module_3_milestone2.py [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+
+
+def main():
+    data = load_data(DATA_PATH)
+    # processed_data = preprocessing(data)
+    # model = training_model(processed_data)
```
Comment (pabaldonedo) [inline src/module_3/module_3_milestone2.py:25]: Remove unused lines

File: src/module_3/module_3_milestone2.py [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+    # save_model(model, model_filename)
+    model = load_model(model_filename)
+    threshold = threshold_calculation(data, model)
+    # with these 3 variables we can predict new values with the model_prediction function
```
Comment (pabaldonedo) [inline src/module_3/module_3_milestone2.py:29]: I do not understand what are you doing here. You should split this in two files one for training and another for inference

File: src/module_3/module_3_milestone2.py [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+
+
+def load_data(data_path: str) -> pd.DataFrame:
```
Comment (pabaldonedo) [inline src/module_3/module_3_milestone2.py:32]: Good to have an auxiliary function to pack the load data

File: src/module_3/module_3_milestone2.py [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+    return pd.read_csv(data_path)
+
+
+def preprocessing(data: pd.DataFrame) -> pd.DataFrame:
+    # We only keep orders with at least 5 items
```
Comment (pabaldonedo) [inline src/module_3/module_3_milestone2.py:37]: This should go in a docstring not in a comment

File: src/module_3/module_3_milestone2.py [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+    model = load_model(model_filename)
+    threshold = threshold_calculation(data, model)
+    # with these 3 variables we can predict new values with the model_prediction function
+
+
+def load_data(data_path: str) -> pd.DataFrame:
+    return pd.read_csv(data_path)
+
+
+def preprocessing(data: pd.DataFrame) -> pd.DataFrame:
```
Comment (pabaldonedo) [inline src/module_3/module_3_milestone2.py:36]: Nice type hinting!

File: src/module_3/module_3_milestone2.py [hunk 8]
```diff
@@ -0,0 +1,104 @@
...
+    # We only keep orders with at least 5 items
+    orders = data["order_id"].unique()
+    orders_more5 = orders[data.groupby("order_id")["outcome"].sum() >= 5]
+    return data.loc[data["order_id"].isin(orders_more5)]
+
+
+def training_model(data: pd.DataFrame) -> make_pipeline:
```
Comment (pabaldonedo) [inline src/module_3/module_3_milestone2.py:43]: [make_pipeline](https://scikit-learn.org/stable/modules/generated/sklearn.pipeline.make_pipeline.html) returns a Pipeline not a make_pipeline

File: src/module_3/module_3_milestone2.py [hunk 9]
```diff
@@ -0,0 +1,104 @@
...
+    model = make_pipeline(
```
Comment (pabaldonedo) [inline src/module_3/module_3_milestone2.py:44]: Nice use of make_pipeline

File: src/module_3/module_3_milestone2.py (hunk 8 above)
Comment (pabaldonedo) [inline src/module_3/module_3_milestone2.py:43]: Minor comment: you are coupling the training with the data split of covariates - target variables. Think that at inference time you want to extract the exact same covariates than here. Coupling both steps makes it harder and brug-prone. I'd suggest to extract somewhere else the split so we can reuse it here and inference time

File: src/module_3/module_3_milestone2.py [hunk 10]
```diff
@@ -0,0 +1,104 @@
...
+        StandardScaler(),
+        linear_model.LogisticRegression(penalty=None, max_iter=250, solver="lbfgs"),
+    )
+
+    return model.fit(data[training_columns], data[label_col])
+
+
+def save_model(model: make_pipeline, file_name: str):
+    joblib.dump(model, file_name)  # file_name must be .pkl
```
Comment (pabaldonedo) [inline src/module_3/module_3_milestone2.py:53]: What if the file_name already exists?

File: src/module_3/module_3_milestone2.py [hunk 11]
```diff
@@ -0,0 +1,104 @@
...
+def training_model(data: pd.DataFrame) -> make_pipeline:
+    model = make_pipeline(
+        StandardScaler(),
+        linear_model.LogisticRegression(penalty=None, max_iter=250, solver="lbfgs"),
+    )
+
+    return model.fit(data[training_columns], data[label_col])
+
+
+def save_model(model: make_pipeline, file_name: str):
```
Comment (pabaldonedo) [inline src/module_3/module_3_milestone2.py:52]: make_pipeline is not a type of any object, it is a function

File: src/module_3/module_3_milestone2.py [hunk 12]
```diff
@@ -0,0 +1,104 @@
...
+    joblib.dump(model, file_name)  # file_name must be .pkl
+
+
+def load_model(model_filename: str) -> make_pipeline:
+    try:
+        loaded_model = joblib.load(model_filename)
+        return loaded_model
+    except FileNotFoundError:
+        print(f"The model file '{model_filename}' was not found.")
+        return None
```
Comment (pabaldonedo) [inline src/module_3/module_3_milestone2.py:62]: Why returning None instead reraising the exception?

File: src/module_3/module_3_milestone2.py [hunk 13]
```diff
@@ -0,0 +1,104 @@
...
+def save_model(model: make_pipeline, file_name: str):
+    joblib.dump(model, file_name)  # file_name must be .pkl
+
+
+def load_model(model_filename: str) -> make_pipeline:
+    try:
+        loaded_model = joblib.load(model_filename)
+        return loaded_model
+    except FileNotFoundError:
+        print(f"The model file '{model_filename}' was not found.")
```
Comment (pabaldonedo) [inline src/module_3/module_3_milestone2.py:61]: Better use logging library than prints

File: src/module_3/module_3_milestone2.py [hunk 14]
```diff
@@ -0,0 +1,104 @@
...
+            precision_recall["Threshold"][
+                (precision_recall["Recall"] >= 0.25)
+                & (precision_recall["Recall"] <= 0.28)
+            ]
+        ),
+        4,
+    )
+
+
+def calculate_probabilities(data: pd.DataFrame, model: make_pipeline) -> np.array:
```
Comment (pabaldonedo) [inline src/module_3/module_3_milestone2.py:90]: minor: these are not probabilities, but a score. I'd use proba to keep naming

File: src/module_3/module_3_milestone2.py [hunk 15]
```diff
@@ -0,0 +1,104 @@
...
+    return model.predict_proba(data[training_columns])[:, 1]
```
Comment (pabaldonedo) [inline src/module_3/module_3_milestone2.py:91]: See, here if we want to do something else than training_columns, but also making some transformations, you would need to sync this with the training method

File: src/module_3/module_3_milestone2.py [hunk 16]
```diff
@@ -0,0 +1,104 @@
...
+        data[label_col], calculate_probabilities(data, model)
+    )
+    precision_recall = pd.DataFrame(
+        {
+            "Precision": precision[: len(precision) - 1],
+            "Recall": recall[: len(recall) - 1],
+            "Threshold": thresholds_precrec,
+        }
+    )
+    return round(
```
Comment (pabaldonedo) [inline src/module_3/module_3_milestone2.py:79]: I do not understand why round and np.mean

**Review by pabaldonedo — APPROVED** [review]
I've left some comments. Read them to apply them for next module.

File: src/module_3/module_3_milestone2.py (hunk 4 above)
Comment (uhernando002) [reply src/module_3/module_3_milestone2.py:29]: In the inference process what should I do?

File: src/module_3/module_3_milestone2.py (hunk 8 above)
Comment (uhernando002) [reply src/module_3/module_3_milestone2.py:43]: I don't understand this comment.

File: src/module_3/module_3_milestone2.py (hunk 15 above)
Comment (uhernando002) [reply src/module_3/module_3_milestone2.py:91]: What type of transformations could be done?

File: src/module_3/module_3_milestone2.py (hunk 16 above)
Comment (uhernando002) [reply src/module_3/module_3_milestone2.py:79]: For a better reading, but as it is an intern parameter, I understand that is unnecessary

File: src/module_3/groceries_model.pkl
Comment (uhernando002) [reply src/module_3/groceries_model.pkl:1]: When you say save the model on disk what should be done?

**Pull request merged**